icu4c/source/common/locid.cpp

```diff
@@ -2770,15 +2796,15 @@ Locale::setKeywordValue(StringPiece keywordName,
             std::string_view script = heap->script;
             std::string_view region = heap->region;
             if (Nest::fits(localeID.length(), language, script, region)) {
-                LocalPointer<Heap> save = payload.release();
-                U_ASSERT(save.isValid());
-                nest = &payload.emplace<Nest>();
+                int32_t variantBegin = heap->ptr->variantBegin;
+                U_ASSERT(variantBegin >= 0);
+                U_ASSERT(static_cast<size_t>(variantBegin) < sizeof Nest::baseName);
+                nest = &payload.emplace<Nest>(script, region, static_cast<uint8_t>(variantBegin));
```

Comment:
I see what you are doing here, and in this special constructor. It's very clever, but it seems misleading to disguise as a constructor what is really a "move Heap fields to Nest fields" function. It would be clearer / much easier to understand to name it appropriately and go back to just `&payload.emplace<Nest>()` and then calling that function.

Reply:
But maybe a different signature might help make it a little bit more clear that the special constructor isn't an ordinary constructor.